pkg/maps/ctmap/ctmap.go

```diff
@@ -508,6 +515,108 @@ func GC(m *Map, filter *GCFilter) int {
+// orphan if it does not have either a corresponding CT entry or an SNAT entry
+// in reverse order.
+//
+// The former case can happen when the CT entry is removed by the LRU eviction
+// which happens when the CT map becomes full. The latter case can happen for
+// the same LRU eviction reasons.
+//
+// PurgeOrphanNATEntries() is triggered by the datapath via the GC signaling
+// mechanism. When the datapath SNAT fails to find free mapping after
+// SNAT_SIGNAL_THRES attempts, it sends the signal via the perf ring buffer.
+// The consumer of the buffer invokes the function.
+//
+// The SNAT is being used for the following cases:
+// 1. By NodePort BPF on an intermediate node before fwd'ing request from outside
+//     to a destination node.
+// 2. A packet from local endpoint sent to outside (BPF-masq).
+// 3. A packet from a host local application (i.e. running in the host netns)
+//    This is needed to prevent SNAT from hijacking such connections.
+// 4. By DSR on a backend node to SNAT responses with service IP+port before
+//    sending to a client.
+//
+// In the case of 1-3, we always create a CT_EGRESS CT entry. This allows the
+// CT GC to remove corresponding SNAT entries. In the case of 4, will create
+// CT_INGRESS CT entry. See the unit test TestOrphanNatGC for more examples.
+//
+// The function only handles 1-3 cases, the 4. case is TODO(brb).
+func PurgeOrphanNATEntries(ctMap *Map) *NatGCStats {
+	if option.Config.NodePortMode == option.NodePortModeDSR ||
+		option.Config.NodePortMode == option.NodePortModeHybrid {
+		return nil
+	}
+
+	natMap := mapInfo[ctMap.mapType].natMap
+	if natMap == nil {
+		return nil
+	}
+	stats := newNatGCStats(natMap)
+
+	cb := func(key bpf.MapKey, value bpf.MapValue) {
+		natKey := key.(*nat.NatKey4)
+		natVal := value.(*nat.NatEntry4)
+
+		if natKey.GetFlags()&tuple.TUPLE_F_IN == 1 { // natKey is r(everse)tuple
+			ctKey := &tuple.TupleKey4Global{TupleKey4: tuple.TupleKey4{
+				// Workaround #5848
+				SourceAddr: natKey.SourceAddr,
+				DestPort:   natKey.SourcePort,
+				DestAddr:   natVal.Addr,
+				SourcePort: natVal.Port,
+				NextHeader: natKey.NextHeader,
+				Flags:      tuple.TUPLE_F_OUT,
+			}}
+			if _, err := ctMap.Lookup(ctKey); errors.Is(err, unix.ENOENT) {
+				// No CT entry is found, so delete SNAT for both original and
+				// reverse flows
+				oNatKey := &nat.NatKey4{TupleKey4Global: tuple.TupleKey4Global{TupleKey4: tuple.TupleKey4{
+					SourceAddr: natVal.Addr,
+					SourcePort: natVal.Port,
+					DestAddr:   natKey.SourceAddr,
+					DestPort:   natKey.SourcePort,
+					NextHeader: natKey.NextHeader,
+					Flags:      tuple.TUPLE_F_OUT,
+				}}}
+				if err := natMap.Delete(natKey); err == nil {
+					stats.IngressDeleted += 1
+				}
+				if err := natMap.Delete(oNatKey); err == nil {
+					stats.EgressDeleted += 1
+				}
+			} else {
+				stats.IngressAlive += 1
+			}
+		} else { // natKey is o(riginal)tuple
```

Comment:
@brb I'm trying to understand the if/else branches here. I understand the flow of CT and SNAT map population as follows,
1. Create a CT entry
2. Create SNAT reply entry (TUPLE_F_IN) ->
3. Create SNAT original dir entry (TUPLE_F_OUT)
My understanding is then for _most_ cases we should have a key for both directions. At some point the CT entry gets deleted and we have an orphaned SNAT entry that needs to be deleted.

At this point we start walking our nat table. If the key has TUPLE_F_IN set AND no CT Entry we will delete both oNatKey and rNatKey. All good as far as I read it.

Now if the key has TUPLE_F_OUT set we use the else branch and check if its rNatKey exists. If it does leave it alone. If not its somehow been orphaned by its pair. Couple questions here.

  - This case where rNatKey of a TUPLE_F_OUT does not exist should be rare correct?
  - It appears to rely on a race condition. Are there are TUPLE_F_* types where it might be more common? 

Now working on my assumption its rare, why bother with the else branch at all? Just check for CT in both TUPLE_F_IN and TUPLE_F_OUT cases and if it does not exist delete both halves of the key. In the race condition handled now in the else branch it would simply fail the delete, no harm done. And by it being rare who cares if we rarely do an extra delete on the map.

Did I miss something? Seems cleaner to just handle the single case and not have an if/else here if possible.

And just to be clear, I'm not arguing it incorrect as is.

Reply:
(cross-linking https://github.com/cilium/cilium/pull/13912#discussion_r518461777 from the master PR which I think is asking about the same base idea)